File: MMDDumper/scripts/native-smoke.mjs
#!/usr/bin/env node
import { spawnSync } from "node:child_process";
import { existsSync, mkdirSync, writeFileSync } from "node:fs";
import { resolve } from "node:path";

const root = resolve(import.meta.dirname, "..");
const buildDir = resolve(root, "out", "native-smoke");
const dllBuildDir = resolve(root, "out", "native-dll-smoke");
const proxyBuildDir = resolve(root, "out", "native-proxy-smoke");
const d3d9BuildDir = resolve(root, "out", "native-d3d9-smoke");
const pluginBuildDir = resolve(root, "out", "native-mmdplugin-smoke");
const mmdExportLib = resolve(root, "lib", "mmd", "MMDExport.lib");
const mmdPluginIncludeDir = resolve(root, "lib", "mmdplugin");
const vsDevCmd = findVsDevCmd();

const cmakeVersion = spawnSync("cmake", ["--version"], { encoding: "utf8" });
if (cmakeVersion.status !== 0) {
  console.log(JSON.stringify({ ok: true, skipped: true, reason: "cmake not found" }, null, 2));
  process.exit(0);
}

const compiler = findCompiler();
if (!compiler && !vsDevCmd) {
  console.log(JSON.stringify({ ok: true, skipped: true, reason: "no C++ compiler found in PATH" }, null, 2));
  process.exit(0);
}

cmakeConfigure(buildDir, ["-DMMD_ORACLE_BUILD_DLL=OFF", "-DMMD_ORACLE_BUILD_MSIMG32_PROXY=OFF"]);
cmakeBuild(buildDir, "mmd_oracle_jsonl_writer_test");
ctest(buildDir);
requireFile(resolve(buildDir, "mmd_oracle_jsonl_writer_test.exe"));

let dllBuilt = false;
if (existsSync(mmdExportLib)) {
  cmakeConfigure(dllBuildDir, ["-DMMD_ORACLE_BUILD_DLL=ON", "-DMMD_ORACLE_BUILD_MSIMG32_PROXY=OFF", `-DMMD_EXPORT_LIB=${mmdExportLib}`]);
  cmakeBuild(dllBuildDir, "mmd_oracle_dumper");
  requireFile(resolve(dllBuildDir, "mmd_oracle_dumper.dll"));
  dllBuilt = true;
}

let pluginBuilt = false;
if (existsSync(mmdExportLib) && existsSync(resolve(mmdPluginIncludeDir, "mmd_plugin.h"))) {
  cmakeConfigure(pluginBuildDir, [
    "-DMMD_ORACLE_BUILD_MMDPLUGIN=ON",
    `-DMMD_EXPORT_LIB=${mmdExportLib}`,
    `-DMMD_PLUGIN_INCLUDE_DIR=${mmdPluginIncludeDir}`,
  ]);
  cmakeBuild(pluginBuildDir, "mmd_oracle_plugin");
  requireFile(resolve(pluginBuildDir, "mmd_oracle_plugin.dll"));
  pluginBuilt = true;
}

cmakeConfigure(proxyBuildDir, ["-DMMD_ORACLE_BUILD_DLL=OFF", "-DMMD_ORACLE_BUILD_MSIMG32_PROXY=ON"]);
cmakeBuild(proxyBuildDir, "msimg32");
requireFile(resolve(proxyBuildDir, "MSIMG32.dll"));

cmakeConfigure(d3d9BuildDir, ["-DMMD_ORACLE_BUILD_DLL=OFF", "-DMMD_ORACLE_BUILD_D3D9_PROXY=ON"]);
cmakeBuild(d3d9BuildDir, "d3d9");
requireFile(resolve(d3d9BuildDir, "d3d9.dll"));

console.log(
  JSON.stringify(
    {
      ok: true,
      buildDir,
      dllBuildDir,
      proxyBuildDir,
      d3d9BuildDir,
      pluginBuildDir,
      dllBuilt,
      dllSkippedReason: dllBuilt ? undefined : "local MMDExport.lib not found",
      pluginBuilt,
      pluginSkippedReason: pluginBuilt ? undefined : "local MMDExport.lib or mmd_plugin.h not found",
      compiler: compiler ?? "Visual Studio Developer Command Prompt",
      compilerCacheExists: existsSync(buildDir),
    },
    null,
    2,
  ),
);

function cmakeConfigure(directory, definitions) {
  run("cmake", ["-S", resolve(root, "native"), "-B", directory, "-G", "Ninja", ...compilerFlag(), ...definitions]);
}

function cmakeBuild(directory, target) {
  run("cmake", ["--build", directory, "--target", target, "--config", "Release"]);
}

function ctest(directory) {
  run("ctest", ["--test-dir", directory, "--output-on-failure", "-C", "Release"]);
}

function run(command, args) {
  const result = vsDevCmd && !compiler ? runInVsDevCmd(command, args) : spawnSync(command, args, { cwd: root, stdio: "inherit", shell: false, timeout: 120000 });
  if (result.error?.code === "ETIMEDOUT") {
    console.error(`${command} timed out`);
    process.exit(124);
  }
  if (result.status !== 0) {
    process.exit(result.status ?? 1);
  }
}

function requireFile(path) {
  if (!existsSync(path)) {
    console.error(`Expected build artifact was not created: ${path}`);
    process.exit(1);
  }
}

function runInVsDevCmd(command, args) {
  mkdirSync(buildDir, { recursive: true });
  const commandLine = [`@echo off`, `call ${quoteCmd(vsDevCmd)} -arch=x64 -host_arch=x64 >nul`, [quoteCmd(command), ...args.map(quoteCmd)].join(" ")].join("\r\n");
  const scriptPath = resolve(buildDir, "run-vs-command.cmd");
  writeFileSync(scriptPath, `${commandLine}\r\n`, "utf8");
  return spawnSync("cmd.exe", ["/d", "/c", scriptPath], {
    cwd: root,
    stdio: "inherit",
    shell: false,
    timeout: 120000,
  });
}

function compilerFlag() {
  return compiler ? ["-DCMAKE_CXX_COMPILER=" + compiler] : [];
}

function findCompiler() {
  for (const command of ["cl", "clang++", "g++", "c++"]) {
    if (commandExists(command)) {
      return command;
    }
  }
  return null;
}

function commandExists(command) {
  const probe = process.platform === "win32" ? spawnSync("where.exe", [command], { encoding: "utf8" }) : spawnSync("command", ["-v", command], { encoding: "utf8", shell: true });
  return probe.status === 0;
}

function findVsDevCmd() {
  if (process.platform !== "win32") {
    return null;
  }
  const candidates = [
    process.env.ProgramFiles ? resolve(process.env.ProgramFiles, "Microsoft Visual Studio", "Installer", "vswhere.exe") : null,
    process.env["ProgramFiles(x86)"] ? resolve(process.env["ProgramFiles(x86)"], "Microsoft Visual Studio", "Installer", "vswhere.exe") : null,
  ].filter(Boolean);

  for (const vswhere of candidates) {
    if (!existsSync(vswhere)) {
      continue;
    }
    const result = spawnSync(vswhere, ["-latest", "-products", "*", "-requires", "Microsoft.VisualStudio.Component.VC.Tools.x86.x64", "-property", "installationPath"], { encoding: "utf8" });
    if (result.status !== 0) {
      continue;
    }
    const installationPath = result.stdout.trim().split(/\r?\n/)[0];
    if (!installationPath) {
      continue;
    }
    const candidate = resolve(installationPath, "Common7", "Tools", "VsDevCmd.bat");
    if (existsSync(candidate)) {
      return candidate;
    }
  }
  return null;
}

function quoteCmd(value) {
  return `"${String(value).replaceAll('"', '""')}"`;
}

File: MMDDumper/scripts/native_smoke.py
#!/usr/bin/env python3
"""Build and test MMDDumper native targets with CMake."""

from __future__ import annotations

import shutil
import subprocess
import os
from pathlib import Path

_GENERATOR_ARGS: tuple[str, ...] = ()

def main() -> int:
    root = Path(__file__).resolve().parents[1]
    if shutil.which("cmake") is None:
        print('{"ok": true, "skipped": true, "reason": "cmake not found"}')
        return 0
    global _GENERATOR_ARGS
    if any(shutil.which(command) for command in ("cl", "clang++", "g++", "c++")):
        _GENERATOR_ARGS = ("-G", "Ninja")
    elif os.name == "nt" and (generator := _visual_studio_generator()) is not None:
        _GENERATOR_ARGS = ("-G", generator, "-A", "x64")
    else:
        print('{"ok": true, "skipped": true, "reason": "no C++ compiler found in PATH"}')
        return 0
    out = root / "out"
    writer = out / "native-smoke"
    _configure_build(root, writer, "mmd_oracle_jsonl_writer_test", "-DMMD_ORACLE_BUILD_DLL=OFF", "-DMMD_ORACLE_BUILD_MSIMG32_PROXY=OFF")
    _run(("ctest", "--test-dir", str(writer), "--output-on-failure", "-C", "Release"), root)
    _require_built(writer, "mmd_oracle_jsonl_writer_test.exe", "mmd_oracle_jsonl_writer_test")

    mmd_export = root / "lib" / "mmd" / "MMDExport.lib"
    dll_built = False
    if mmd_export.is_file():
        dll = out / "native-dll-smoke"
        _configure_build(root, dll, "mmd_oracle_dumper", "-DMMD_ORACLE_BUILD_DLL=ON", "-DMMD_ORACLE_BUILD_MSIMG32_PROXY=OFF", f"-DMMD_EXPORT_LIB={mmd_export}")
        _require_built(dll, "mmd_oracle_dumper.dll")
        dll_built = True
    proxy = out / "native-proxy-smoke"
    _configure_build(root, proxy, "msimg32", "-DMMD_ORACLE_BUILD_DLL=OFF", "-DMMD_ORACLE_BUILD_MSIMG32_PROXY=ON")
    _require_built(proxy, "MSIMG32.dll")
    d3d9 = out / "native-d3d9-smoke"
    _configure_build(root, d3d9, "d3d9", "-DMMD_ORACLE_BUILD_DLL=OFF", "-DMMD_ORACLE_BUILD_D3D9_PROXY=ON")
    _require_built(d3d9, "d3d9.dll")
    print(f'{{"ok": true, "dllBuilt": {str(dll_built).lower()}, "writer": "{writer}"}}')
    return 0


def _configure_build(root: Path, build: Path, target: str, *definitions: str) -> None:
    cache = build / "CMakeCache.txt"
    if cache.is_file():
        cache_text = cache.read_text(encoding="utf-8", errors="replace")
        requested_generator = "Visual Studio" if any("Visual Studio" in arg for arg in _GENERATOR_ARGS) else "Ninja"
        if "MMDDumper/native" not in cache_text or requested_generator not in cache_text:
            shutil.rmtree(build)
    _run(("cmake", "-S", str(root / "native"), "-B", str(build), *_GENERATOR_ARGS, *definitions), root)
    _run(("cmake", "--build", str(build), "--target", target, "--config", "Release"), root)


def _run(command: tuple[str, ...], cwd: Path) -> None:
    try:
        completed = subprocess.run(command, cwd=cwd, check=False, timeout=120)
    except subprocess.TimeoutExpired:
        print("native smoke command timed out after 120 seconds", flush=True)
        raise SystemExit(124)
    if completed.returncode != 0:
        raise SystemExit(completed.returncode)


def _require_built(build: Path, *names: str) -> None:
    paths = tuple(directory / name for directory in (build, build / "Release") for name in names)
    if not any(path.is_file() for path in paths):
        raise FileNotFoundError("expected native artifact was not created: " + ", ".join(str(path) for path in paths))


def _visual_studio_generator() -> str | None:
    candidates = []
    for variable in ("ProgramFiles", "ProgramFiles(x86)"):
        value = os.environ.get(variable)
        if value:
            candidates.append(Path(value) / "Microsoft Visual Studio" / "Installer" / "vswhere.exe")
    for path in candidates:
        if not path.is_file():
            continue
        result = subprocess.run((str(path), "-latest", "-products", "*", "-requires", "Microsoft.VisualStudio.Component.VC.Tools.x86.x64", "-property", "installationPath"), capture_output=True, check=False, timeout=10, text=True)
        if result.returncode != 0:
            continue
        major = result.stdout.strip().split("\\")[-2] if result.stdout.strip() else ""
        if major == "18":
            return "Visual Studio 18 2026"
        if major == "17":
            return "Visual Studio 17 2022"
        return "Visual Studio 17 2022"
    return None


if __name__ == "__main__":
    raise SystemExit(main())
